packages/server/src/lockers/GCSLock.ts

```diff
@@ -0,0 +1,120 @@
+import {RequestRelease} from '@tus/utils'
+import {Bucket} from '@google-cloud/storage'
+import GCSLockFile, {GCSLockFileMetadata} from './GCSLockFile'
+
+/**
+ * Handles interaction with a lock.
+ */
+class GCSLock {
+  protected resourceId: string
+  protected file: GCSLockFile
+  protected ttl: number
+  protected watchInterval: number
+  protected watcher: NodeJS.Timeout | undefined
+
+  constructor(
+    resourceId: string,
+    lockBucket: Bucket,
+    ttl: number,
+    watchInterval: number
+  ) {
+    this.resourceId = resourceId
+    this.file = new GCSLockFile(lockBucket, `${resourceId}.lock`)
+    this.ttl = ttl
+    this.watchInterval = watchInterval
+  }
+
+  /**
+   * Try to create the lockfile and start the watcher. If lock is already taken, requests for release and returns FALSE.
+   */
+  public async take(cancelHandler: RequestRelease): Promise<boolean> {
+    try {
+      //Try to create lock file
+      const exp = Date.now() + this.ttl
+      await this.file.create(exp)
+
+      //Lock acquired, start watcher
+      this.startWatcher(cancelHandler)
+
+      return true
+    } catch (err) {
+      //Probably lock is already taken
+      const isHealthy = await this.insureHealth()
+
+      if (!isHealthy) {
+        //Lock is not healthy, restart the process
+        return await this.take(cancelHandler)
+      } else {
+        //Lock is still healthy, request release
+        await this.file.requestRelease()
+
+        return false
+      }
+    }
+  }
+
+  /**
+   * Release the lock - clear watcher and delete the file.
+   */
+  public async release() {
```

Comment:
Do we need some internal mutex to ensure that deleting a lock and extending it cannot run simultaneously and thus interfere with each other? Async requests can run concurrently.

Reply:
AFAIK GCS should already take care of concurrent access for us.